packages/@aws-cdk/aws-lambda/lib/function-base.ts

```diff
@@ -566,6 +568,11 @@ export abstract class FunctionBase extends Resource implements IFunction, ec2.IC
       return (principal as iam.ArnPrincipal).arn;
     }
 
+    if ('organizationId' in principal) {
```

Comment:
This feels nasty, and should not be necessary? `OrganizationPrincipals` fragment looks like this:

```ts
  public get policyFragment(): PrincipalPolicyFragment {
    return new PrincipalPolicyFragment(
      { AWS: ['*'] },
      { StringEquals: { 'aws:PrincipalOrgID': this.organizationId } },
    );
  }
```

So it should already fall properly into the cases below, where we parse the output of `policyFragment` and parse the `AWS` key out?

(Honestly -- that should be the only thing we do. The duck typing here above is 🤮)

Reply:
i'm down to move it out, but the case below has the comment 

```
// The principal cannot have conditions and must have a single { AWS: [arn] } entry.
```

`OrganizationPrincipal` has the condition `StringEquals`, so it fails. @rix0rrr since you were the one who introduced that comment, can you explain why the principal cannot have conditions? maybe we can loosen the logic to allow for something like `OrganizationPrincipal` to be parsed, but until then, the duck typing stays :).

Reply:
```ts
    // Try a best-effort approach to support simple principals that are not any of the predefined
    // classes, but are simple enough that they will fit into the Permission model. Main target
    // here: imported Roles, Users, Groups.
    //
    // The principal cannot have conditions and must have a single { AWS: [arn] } entry.
    const json = principal.policyFragment.principalJson;
    if (Object.keys(principal.policyFragment.conditions).length === 0 && json.AWS) {
      if (typeof json.AWS === 'string') { return json.AWS; }
      if (Array.isArray(json.AWS) && json.AWS.length === 1 && typeof json.AWS[0] === 'string') {
        return json.AWS[0];
      }
    }
```

the parsing logic, for reference.

Reply:
In that case, I'd still prefer to see something like:

```ts
const stringEquals = matchSingleKey('StringEquals', principal.policyFragment.conditions));
if (stringEquals) {
  const orgId = matchSingleKey('aws:PrincipalOrgID', stringEquals);
  if (orgId) {
    return '*';
  }
}
```

Right?

What I'm trying to get at is that we look at the existing AWS principal and conditions, and parse and convert those to Lambda Permissions.

*Instead of*, what the code is historically doing, is looking at what *classes* you used and using "inside" knowledge of what those classes represent to generate the right output.

Reply:
I've done this for org principal. I wanted to refactor the others to follow the same approach, but I couldn't get it to work. The approach works for org principal because we can identify the org principal from its policy fragment. That isn't the case for something like `ArnPrincipal` and `AccountPrincipal`, since both their policy fragments have just an `arn` value.